uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Import/Parser/BaseTest.php

```diff
@@ -0,0 +1,288 @@
+      'period_id' => $period->id
+    ]);
+
+    $entitlementBalanceChange = 3;
+    $this->createLeaveBalanceChange($periodEntitlement->id, $entitlementBalanceChange);
+
+    HRJobContractFabricator::fabricate(
+      ['contact_id' => $periodEntitlement->contact_id],
+      ['period_start_date' => '2016-01-01']
+    );
+
+    WorkPatternFabricator::fabricateWithA40HourWorkWeek(['is_default' => 1]);
+
+    //This will create a balance change of 4 since the value for total_qty
+    //is 4
+    $row1 = [
+      'contact_id' => $periodEntitlement->contact_id,
+      'absence_id' => 1,
+      'type_id' => $absenceType->id,
+      'absence_date' => '2016-01-01',
+      'qty' => 4,
+      'start_date' => '2016-01-01',
+      'end_date' => '2016-01-01',
+      'total_qty' => 4,
+      'status_id' => 1,
+      'request_type' => LeaveRequest::REQUEST_TYPE_TOIL,
+    ];
+
+    $fields = array_keys($row1);
+    $importObject = $this->getImportObject($fields);
+    //Since the request type is TOIL, balance change validation will not
+    //be taken into account.
+    $importObject->validateLeaveParams($row1);
+  }
+
+  public function testValidateLeaveParamsDoesNotThrowsExceptionWhenBalanceIsGreaterThanEntitlementBalanceWhenAllowOveruseFalseAndRequestIsCancelled() {
+    $period = AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-01-01'),
+      'end_date'   => CRM_Utils_Date::processDate('2016-12-31'),
+    ]);
+
+    $absenceType = AbsenceTypeFabricator::fabricate([
+      'title' => 'Type 1',
+      'allow_overuse' => 0
+    ]);
+
+    $periodEntitlement = LeavePeriodEntitlementFabricator::fabricate([
+      'type_id' => $absenceType->id,
+      'contact_id' => 1,
+      'period_id' => $period->id
+    ]);
+
+    $entitlementBalanceChange = 3;
+    $this->createLeaveBalanceChange($periodEntitlement->id, $entitlementBalanceChange);
+
+    HRJobContractFabricator::fabricate(
+      ['contact_id' => $periodEntitlement->contact_id],
+      ['period_start_date' => '2016-01-01']
+    );
+
+    WorkPatternFabricator::fabricateWithA40HourWorkWeek(['is_default' => 1]);
+    LeaveRequest::getCancelledStatuses();
+    //This will create a balance change of 4 since the value for total_qty
+    //is 4
+    $row1 = [
+      'contact_id' => $periodEntitlement->contact_id,
+      'absence_id' => 1,
+      'type_id' => $absenceType->id,
+      'absence_date' => '2016-01-01',
+      'qty' => 4,
+      'start_date' => '2016-01-01',
+      'end_date' => '2016-01-01',
+      'total_qty' => 4,
+      'request_type' => LeaveRequest::REQUEST_TYPE_LEAVE,
+    ];
+
+    $fields = array_keys($row1);
+    $importObject = $this->getImportObject($fields);
+    //Since the request type is TOIL, balance change validation will not
+    //be taken into account.
+    foreach(LeaveRequest::getCancelledStatuses() as $status) {
+      $row1['status_id'] = $status;
+      $importObject->validateLeaveParams($row1);
+    }
+  }
```

Comment:
Instead of doing validations for the `validateLeaveParams()` params method, wouldn't it be better to test if the LeaveRequest can or cannot be imported on those scenarios and keep the validation method private?

Reply:
If we test the using the import method of the `CRM_HRLeaveAndAbsences_Import_Parser_Base` class, the exceptions are already caught there and all it would return is a `CRM_Import_Parser::ERROR` error. Would it be fine to just test for that?

Reply:
Yeah. One of the problems of testing only the `validateLeaveParams()` method is that there's no way to guarantee that this method will actually be called when you run the import. So I would say it's more important to test that we'll get an error trying to import an invalid LeaveRequest than to test the validation in isolation.